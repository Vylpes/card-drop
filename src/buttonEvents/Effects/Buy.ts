import {ButtonInteraction} from "discord.js";
import AppLogger from "../../client/appLogger";
import EffectHelper from "../../helpers/EffectHelper";
import EmbedColours from "../../constants/EmbedColours";
import User from "../../database/entities/app/User";
import {EffectDetails} from "../../constants/EffectDetails";

export default class Buy {
    public static async Execute(interaction: ButtonInteraction) {
        const subaction = interaction.customId.split(" ")[2];

        switch (subaction) {
            case "confirm":
                await this.Confirm(interaction);
                break;
            case "cancel":
                await this.Cancel(interaction);
                break;
            default:
                AppLogger.LogError("Buy", `Unknown subaction, effects ${subaction}`);
        }
    }

    private static async Confirm(interaction: ButtonInteraction) {
        const id = interaction.customId.split(" ")[3];
        const quantity = interaction.customId.split(" ")[4];

        if (!id || !quantity) {
            AppLogger.LogError("Buy Confirm", "Not enough parameters");
            return;
        }
        
        const effectDetail = EffectDetails.get(id);

        if (!effectDetail) {
            AppLogger.LogError("Buy Confirm", "Effect detail not found!");
            return;
        }

        const quantityNumber = Number(quantity);
        
        if (!quantityNumber || quantityNumber < 1) {
            AppLogger.LogError("Buy Confirm", "Invalid number");
            return;
        }

        const totalCost = effectDetail.cost * quantityNumber;

        const user = await User.FetchOneById(User, interaction.user.id);

        if (!user) {
            AppLogger.LogError("Buy Confirm", "Unable to find user");
            return;
        }

        if (user.Currency < totalCost) {
            interaction.reply(`You don't have enough currency to buy this! You have \`${user.Currency} Currency\` and need \`${totalCost} Currency\`!`);
            return;
        }

        user.RemoveCurrency(totalCost);
        await user.Save(User, user);

        await EffectHelper.AddEffectToUserInventory(interaction.user.id, id, quantityNumber);

        const generatedEmbed = await EffectHelper.GenerateEffectBuyEmbed(interaction.user.id, id, quantityNumber, true);

        if (typeof generatedEmbed == "string") {
            await interaction.reply(generatedEmbed);
            return;
        }

        generatedEmbed.embed.setColor(EmbedColours.Success);
        generatedEmbed.embed.setFooter({ text: "Purchased" });

        await interaction.update({
            embeds: [ generatedEmbed.embed ],
            components: [ generatedEmbed.row ],
        });
    }

    private static async Cancel(interaction: ButtonInteraction) {
    }
}
