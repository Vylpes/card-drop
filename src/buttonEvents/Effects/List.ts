import { ButtonInteraction } from "discord.js";
import EffectHelper from "../../helpers/EffectHelper";

export default async function List(interaction: ButtonInteraction) {
    const pageOption = interaction.customId.split(" ")[2];

    const page = Number(pageOption);

    if (!page) {
        await interaction.reply("Page option is not a valid number");
        return;
    }

    const result = await EffectHelper.GenerateEffectEmbed(interaction.user.id, page);

    await interaction.update({
        embeds: [ result.embed ],
        components: [ result.row ],
    });
}