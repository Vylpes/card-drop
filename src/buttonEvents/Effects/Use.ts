import { ActionRowBuilder, ButtonBuilder, ButtonInteraction, ButtonStyle, EmbedBuilder } from "discord.js";
import { EffectDetails } from "../../constants/EffectDetails";
import EffectHelper from "../../helpers/EffectHelper";
import EmbedColours from "../../constants/EmbedColours";
import TimeLengthInput from "../../helpers/TimeLengthInput";
import AppLogger from "../../client/appLogger";

export default async function Use(interaction: ButtonInteraction) {
    const subaction = interaction.customId.split(" ")[2];

    switch (subaction) {
        case "confirm":
            await UseConfirm(interaction);
            break;
        case "cancel":
            await UseCancel(interaction);
            break;
    }
}

export async function UseConfirm(interaction: ButtonInteraction) {
    const id = interaction.customId.split(" ")[3];

    const effectDetail = EffectDetails.get(id);

    if (!effectDetail) {
        AppLogger.LogError("Button/Effects/Use", `Effect not found, ${id}`);

        await interaction.reply("Effect not found in system!");
        return;
    }

    const now = new Date();

    const whenExpires = new Date(now.getTime() + effectDetail.duration);

    const result = await EffectHelper.UseEffect(interaction.user.id, id, whenExpires);

    if (!result) {
        await interaction.reply("Unable to use effect! Please make sure you have it in your inventory and is not on cooldown");
        return;
    }

    const embed = new EmbedBuilder()
        .setTitle("Effect Used")
        .setDescription("You now have an active effect!")
        .setColor(EmbedColours.Green)
        .addFields([
            {
                name: "Effect",
                value: effectDetail.friendlyName,
                inline: true,
            },
            {
                name: "Expires",
                value: `<t:${Math.round(whenExpires.getTime() / 1000)}:f>`,
                inline: true,
            },
        ]);

    const row = new ActionRowBuilder<ButtonBuilder>()
        .addComponents([
            new ButtonBuilder()
                .setLabel("Confirm")
                .setCustomId(`effects use confirm ${effectDetail.id}`)
                .setStyle(ButtonStyle.Primary)
                .setDisabled(true),
            new ButtonBuilder()
                .setLabel("Cancel")
                .setCustomId(`effects use cancel ${effectDetail.id}`)
                .setStyle(ButtonStyle.Danger)
                .setDisabled(true),
        ]);

    await interaction.update({
        embeds: [ embed ],
        components: [ row ],
    });
}

export async function UseCancel(interaction: ButtonInteraction) {
    const id = interaction.customId.split(" ")[3];

    const effectDetail = EffectDetails.get(id);

    if (!effectDetail) {
        AppLogger.LogError("Button/Effects/Cancel", `Effect not found, ${id}`);

        await interaction.reply("Effect not found in system!");
        return;
    }

    const timeLengthInput = TimeLengthInput.ConvertFromMilliseconds(effectDetail.duration);

    const embed = new EmbedBuilder()
        .setTitle("Effect Use Cancelled")
        .setDescription("The effect from your inventory has not been used")
        .setColor(EmbedColours.Grey)
        .addFields([
            {
                name: "Effect",
                value: effectDetail.friendlyName,
                inline: true,
            },
            {
                name: "Expires",
                value: timeLengthInput.GetLengthShort(),
                inline: true,
            },
        ]);

    const row = new ActionRowBuilder<ButtonBuilder>()
        .addComponents([
            new ButtonBuilder()
                .setLabel("Confirm")
                .setCustomId(`effects use confirm ${effectDetail.id}`)
                .setStyle(ButtonStyle.Primary)
                .setDisabled(true),
            new ButtonBuilder()
                .setLabel("Cancel")
                .setCustomId(`effects use cancel ${effectDetail.id}`)
                .setStyle(ButtonStyle.Danger)
                .setDisabled(true),
        ]);

    await interaction.update({
        embeds: [ embed ],
        components: [ row ],
    });
}