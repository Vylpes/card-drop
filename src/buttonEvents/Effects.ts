import { ButtonInteraction } from "discord.js";
import { ButtonEvent } from "../type/buttonEvent";
import List from "./Effects/List";
import Use from "./Effects/Use";
import AppLogger from "../client/appLogger";
import Buy from "./Effects/Buy";

export default class Effects extends ButtonEvent {
    public override async execute(interaction: ButtonInteraction) {
        const action = interaction.customId.split(" ")[1];

        switch (action) {
            case "list":
                await List(interaction);
                break;
            case "use":
                await Use.Execute(interaction);
                break;
            case "buy":
                await Buy.Execute(interaction);
                break;
            default:
                AppLogger.LogError("Buttons/Effects", `Unknown action, ${action}`);
        }
    }
}
