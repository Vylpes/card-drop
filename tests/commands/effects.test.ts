import Effects from "../../src/commands/effects";

describe("constructor", () => {
    let effects: Effects;

    beforeEach(() => {
        effects = new Effects();
    });

    test("EXPECT CommandBuilder to be defined", () => {
        expect(effects.CommandBuilder).toMatchSnapshot();
    });
});

describe("execute", () => {
    describe("GIVEN interaction is not a chat input command", () => {
        let interaction: any;

        let listSpy: any;

        beforeEach(async () => {
            interaction = {
                isChatInputCommand: jest.fn().mockReturnValue(false),
            };

            const effects = new Effects();

            listSpy = jest.spyOn(effects as any, "List");

            await effects.execute(interaction);
        });

        test("EXPECT isChatInputCommand to have been called", () => {
            expect(interaction.isChatInputCommand).toHaveBeenCalledTimes(1);
        });

        test("EXPECT nothing to happen", () => {
            expect(listSpy).not.toHaveBeenCalled();
        });
    });

    describe("GIVEN subcommand is list", () => {
        test.todo("EXPECT list function to be called");
    });
});

describe("List", () => {
    describe("GIVEN page option is supplied", () => {
        describe("AND page is a valid number", () => {
            test.todo("EXPECT EffectHelper.GenerateEffectEmbed to have been called with page");

            test.todo("EXPECT interaction to have been replied");
        });

        describe("AND page is not a valid number", () => {
            test.todo("EXPECT EffectHelper.GenerateEffectEmbed to have been called with page of 1");
        });
    });

    describe("GIVEN page option is not supplied", () => {
        test.todo("EXPECT EffectHelper.GenerateEffectEmbed to have been called with a page of 1");
    });
});
