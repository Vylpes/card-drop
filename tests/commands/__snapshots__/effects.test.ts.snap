// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EXPECT CommandBuilder to be defined 1`] = `
{
  "contexts": undefined,
  "default_member_permissions": undefined,
  "default_permission": undefined,
  "description": "Effects",
  "description_localizations": undefined,
  "dm_permission": undefined,
  "integration_types": undefined,
  "name": "effects",
  "name_localizations": undefined,
  "nsfw": undefined,
  "options": [
    {
      "description": "List all effects I have",
      "description_localizations": undefined,
      "name": "list",
      "name_localizations": undefined,
      "options": [
        {
          "autocomplete": undefined,
          "choices": undefined,
          "description": "The page number",
          "description_localizations": undefined,
          "max_value": undefined,
          "min_value": 1,
          "name": "page",
          "name_localizations": undefined,
          "required": false,
          "type": 10,
        },
      ],
      "type": 1,
    },
    {
      "description": "Use an effect in your inventory",
      "description_localizations": undefined,
      "name": "use",
      "name_localizations": undefined,
      "options": [
        {
          "autocomplete": undefined,
          "choices": [
            {
              "name": "Unclaimed Chance Up",
              "name_localizations": undefined,
              "value": "unclaimed",
            },
          ],
          "description": "The effect id to use",
          "description_localizations": undefined,
          "max_length": undefined,
          "min_length": undefined,
          "name": "id",
          "name_localizations": undefined,
          "required": true,
          "type": 3,
        },
      ],
      "type": 1,
    },
    {
      "description": "Buy more effects",
      "description_localizations": undefined,
      "name": "buy",
      "name_localizations": undefined,
      "options": [
        {
          "autocomplete": undefined,
          "choices": [
            {
              "name": "Unclaimed Chance Up",
              "name_localizations": undefined,
              "value": "unclaimed",
            },
          ],
          "description": "The effect id to buy",
          "description_localizations": undefined,
          "max_length": undefined,
          "min_length": undefined,
          "name": "id",
          "name_localizations": undefined,
          "required": true,
          "type": 3,
        },
        {
          "autocomplete": undefined,
          "choices": undefined,
          "description": "The amount to buy",
          "description_localizations": undefined,
          "max_value": undefined,
          "min_value": 1,
          "name": "quantity",
          "name_localizations": undefined,
          "required": false,
          "type": 10,
        },
      ],
      "type": 1,
    },
  ],
  "type": 1,
}
`;
